File: viewer/backend/schemas/_inspection.py
"""Model inspection schemas."""

from __future__ import annotations

from typing import Literal

from pydantic import Field

from viewer.backend.schemas._base import (
    ApiResponseModel,
    ConfigOverrides,
    JsonObject,
    JsonValue,
)


class GraphConfigFieldResponse(ApiResponseModel):
    key: str
    value: JsonValue


class GraphConfigResponse(ApiResponseModel):
    typeName: str
    fields: list[GraphConfigFieldResponse]


class GraphNodeResponse(ApiResponseModel):
    id: str
    label: str
    typeName: str
    path: str
    graphRole: Literal["architecture", "internal", "runtime"]
    parameterCount: int
    parameterSizeBytes: int
    details: JsonObject
    config: GraphConfigResponse | None


class GraphEdgeResponse(ApiResponseModel):
    id: str
    source: str
    target: str


class InspectRequest(ApiResponseModel):
    modelType: str
    model: str
    preset: str
    overrides: ConfigOverrides = Field(default_factory=dict)
    dataset: str | None = None


class InspectResponse(ApiResponseModel):
    modelType: str
    model: str
    preset: str
    parameterCount: int
    parameterSizeBytes: int
    nodes: list[GraphNodeResponse]
    edges: list[GraphEdgeResponse]


class OperationGraphNodeResponse(ApiResponseModel):
    id: str
    label: str
    opKind: Literal[
        "placeholder",
        "call_function",
        "call_module",
        "call_method",
        "get_attr",
        "output",
    ]
    target: str
    modulePath: str | None = None
    groupId: str | None = None
    details: JsonObject = Field(default_factory=dict)


class OperationGraphEdgeResponse(ApiResponseModel):
    id: str
    source: str
    target: str


class OperationGraphResponse(ApiResponseModel):
    modelType: str
    model: str
    preset: str
    source: Literal["torch-export"]
    status: Literal["ok", "unsupported"]
    nodes: list[OperationGraphNodeResponse]
    edges: list[OperationGraphEdgeResponse]
    warnings: list[str] = Field(default_factory=list)

File: viewer/frontend/src/lib/api/inspection.ts
import { z } from "zod";

import { requestJson } from "@/lib/api/client";
import {
  jsonObjectSchema,
  jsonValueSchema,
  type ConfigOverrides,
} from "@/lib/api/schemas";

export const graphConfigSchema = z.object({
  typeName: z.string(),
  fields: z.array(
    z.object({
      key: z.string(),
      value: jsonValueSchema,
    }),
  ),
});

export const graphNodeSchema = z.object({
  id: z.string(),
  label: z.string(),
  typeName: z.string(),
  path: z.string(),
  graphRole: z.enum(["architecture", "internal", "runtime"]),
  parameterCount: z.number(),
  parameterSizeBytes: z.number().default(0),
  details: jsonObjectSchema,
  config: graphConfigSchema.nullable(),
});

export const graphEdgeSchema = z.object({
  id: z.string(),
  source: z.string(),
  target: z.string(),
});

export const inspectResponseSchema = z.object({
  modelType: z.string(),
  model: z.string(),
  preset: z.string(),
  parameterCount: z.number(),
  parameterSizeBytes: z.number().default(0),
  nodes: z.array(graphNodeSchema),
  edges: z.array(graphEdgeSchema),
});

export const operationGraphNodeSchema = z
  .object({
    id: z.string(),
    label: z.string(),
    opKind: z.enum([
      "placeholder",
      "call_function",
      "call_module",
      "call_method",
      "get_attr",
      "output",
    ]),
    target: z.string(),
    modulePath: z.string().nullable(),
    groupId: z.string().nullable(),
    details: jsonObjectSchema,
  })
  .strict();

export const operationGraphEdgeSchema = z
  .object({
    id: z.string(),
    source: z.string(),
    target: z.string(),
  })
  .strict();

export const operationGraphResponseSchema = z
  .object({
    model: z.string(),
    modelType: z.string(),
    preset: z.string(),
    source: z.literal("torch-export"),
    status: z.enum(["ok", "unsupported"]),
    nodes: z.array(operationGraphNodeSchema),
    edges: z.array(operationGraphEdgeSchema),
    warnings: z.array(z.string()),
  })
  .strict();

export type GraphConfig = z.infer<typeof graphConfigSchema>;
export type GraphNode = z.infer<typeof graphNodeSchema>;
export type GraphEdge = z.infer<typeof graphEdgeSchema>;
export type InspectResponse = z.infer<typeof inspectResponseSchema>;
export type OperationGraphNode = z.infer<typeof operationGraphNodeSchema>;
export type OperationGraphEdge = z.infer<typeof operationGraphEdgeSchema>;
export type OperationGraphResponse = z.infer<typeof operationGraphResponseSchema>;

export function inspectModel(input: {
  modelType: string;
  model: string;
  preset: string;
  overrides: ConfigOverrides;
  dataset?: string;
}) {
  return requestJson("/inspect", inspectResponseSchema, {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export function inspectOperationGraph(input: {
  modelType: string;
  model: string;
  preset: string;
  overrides: ConfigOverrides;
  dataset?: string;
}) {
  return requestJson("/inspect/operation-graph", operationGraphResponseSchema, {
    method: "POST",
    body: JSON.stringify(input),
  });
}
